src/Connection.php

```diff
@@ -1617,27 +1624,71 @@ public function createQueryBuilder(): QueryBuilder
     }
 
     /**
-     * Returns a human-readable representation of an array of parameters.
+     * Returns a human-readable representation of a parameter or array of parameters.
      * This properly handles binary data by returning a hex representation.
      *
-     * @param mixed[] $params
+     * @param mixed $param
      */
-    private function formatParameters(array $params): string
+    private function formatParameter($param): string
     {
-        return '[' . implode(', ', array_map(static function ($param): string {
-            if (is_resource($param)) {
-                return (string) $param;
+        if ($param === null) {
+            return 'null';
+        }
+
+        if (is_scalar($param)) {
+            if (is_string($param) && (! $this->isValidUTF8($param) || $this->containsNonPrintableASCIIChars($param))) {
```

Comment:
After sleeping on it, I'm not sure if DBAL should be in the business of formatting prepared statement parameters at all.

The responsibility of the exception is to tell what went wrong by its error message and encapsulate the context for future handling or troubleshooting. By imprinting prepared statement parameters into the message, it may do a disservice: it makes it impossible to log the error message w/o logging the potentially sensitive information stored in the exception context (e.g. tokens, password hashes, personally identifiable information, etc).

Let's try to explore the approach of storing the parameters with the exception w/o making them part of the message. This way, the handling part will be free to process them in an appropriate way (e.g. ignore all parameters except for passwords and tokens) or ignore them.

Reply:
Prototype available in #4387!